mgr/dashboard: Added graphs to visualise perf counter time sequences

Saumay-Agrawal opened the pull request:
- Tables can now be sorted on the basis of space usage, r/w activities, daemon status etc.
- Tables can now be searched for specific details

Signed-off-by: Saumay Agrawal <saumay.agrawal@gmail.com>

tchaikov commented:
@Saumay-Agrawal  needs rebase and a screenshot would be helpful.

tchaikov commented on src/pybind/mgr/dashboard/osd_perf.html:
```diff
@@ -121,7 +121,7 @@
 <section class="content">
 
-    <table>
+    <!-- <table>
```
why you are commenting out this table?

tchaikov replied:
@Saumay-Agrawal why you are commenting out these two heapmap and adding two new ones?

tchaikov commented:
![rw-perf](https://user-images.githubusercontent.com/59071/28707936-dcffe5f2-73ac-11e7-9f2b-058e8c896426.png)

@Saumay-Agrawal could you explain a little bit how this diagram helps us understand the performance of the OSD ?

jcsp commented:
@tchaikov the osd read/write latency bits are not new in this PR -- they were a rough thing that I added a while ago.

Saumay-Agrawal commented:
The changes added in this commit are as follows:

**Addition of get methods to module.py**

- Added a method get_counter_allosd(self, path) which consolidates the time sequences for the perf counter specified as the argument 'path', given by get_counter() method for all the osds in a single JSON object containing {osd_id: time_sequence} pairs.
- Added a method get_counter_stats(self, path) which uses the data collected by get_counter_allosd() and process it to return the statistical data (minimum, maximum, average and stdevs) about the perf counter specified as the argument 'path'.
- Added perf_graphs_prototypes(self, path) which shows the various sample graphs of the given perf counter. It exposes the webpage perf_graphs_prototypes.html at localhost:41000/perf_graph_prototypes/{perf_counter}. These graphs serves the purpose of a general machinery, and will make implementation of performance graphs easier for future developers.
- Added get_perf_graph_data(self) and osd_perf_graph(self) to expose the webpage containing read and write latency perf counters on the dashboard.

![general_prototypes_snapshot](https://user-images.githubusercontent.com/24296199/29542075-4080114a-86f6-11e7-92b2-bdd8832e303b.png)


**General UI features of the graphs**

- The perf counter values for all OSDs are represented by thin white lines.
- The minimum and maximum perf counter values are respresented by thick green colored lines.
- The average perf counter values are represented by thick orange colored lines.
- The standard deviations on both sides of the average value are represented by thick black lines.
- On hovering the mouse pointer on points plotted on the graph, exact value for that dataset is shown in a tooltip along with its label. For eg. OSD id can be identified on seeing a spike on white line by hovering over it.
- If two or more datasets intersect at the same point, i.e. they have same value for a time instant, then the tool tip lists all the values along with their labels.

![latency_perf_graph_snapshots](https://user-images.githubusercontent.com/24296199/29542086-4cc674c6-86f6-11e7-8c19-46393344dc15.png)


**OSD Performance Graph (Webpage visualizing Latencies)**
- This page is linked to the Cluster/OSDs page on the dashboard.
- This page contains four graphs. Two graphs show the minimum, maximum and average values for the read and write latencies each. And the other two graphs show the distribution of read and write latencies in terms of process and prepare latencies.

tchaikov commented on src/pybind/mgr/dashboard/osd_perf.html:
```diff
@@ -121,20 +121,6 @@
 
 <section class="content">
 
-    <!-- <table>
```
please squash this change into 65bd5188c00a8dd68823fb5db2d85c5a573e75aa

tchaikov commented on src/pybind/mgr/dashboard/module.py:
```diff
@@ -751,20 +752,110 @@ def get_perf_schema(self, **args):
 
                 return result
 
+            @cherrypy.expose
```
why would you want to expose this method?

Saumay-Agrawal replied:
This function is no longer there in the final commit. Again, I would like your opinion on squashing all the changes related to graphs and TS in a single commit.

tchaikov commented on src/pybind/mgr/dashboard/module.py:
```diff
@@ -751,20 +752,110 @@ def get_perf_schema(self, **args):
+                toplevel_data = self._toplevel_data()
+
+                content_data = {
+                    "avg" : self.get_counter_avg(path),
```
no need to `get_counter_data()` repeatedly, please store it locally. and get the avg and stddev in a single run.

Saumay-Agrawal replied:
I admit you are right. I realised the same while I was working on the graphs, so I did replace the entire set of separate methods into a single method get_counter_stats() which stores the data of get_counter_allosd() one time, and then processes it to return the needed values of average, stdevs etc.

tchaikov commented on src/pybind/mgr/dashboard/module.py:
```diff
@@ -751,20 +752,110 @@ def get_perf_schema(self, **args):
+
+                return result
+
+            #Method to find the average of consolidated data
```
s/Method to /

Saumay-Agrawal replied:
It is fixed in the latest commit. Perhaps, I should squash all the changes related to graphs in a single commit.

tchaikov replied:
@Saumay-Agrawal commits should be self-contained and be atomic. please do so.

tchaikov commented on src/pybind/mgr/dashboard/perf_graphs.html:
```diff
@@ -0,0 +1,168 @@
+    var avg_values = convert_timeseries(content_data["avg"]);
+    var stdev_values = convert_timeseries(content_data["stdev"]);
+
+	var minmax = document.getElementById("minmax-chart");
```
could you be more consistent when using space and tab for indent?

tchaikov commented on src/pybind/mgr/dashboard/perf_graphs.html:
```diff
@@ -0,0 +1,168 @@
+				fillColor: "rgba(142, 94, 162, 0.2)",
+			}
+			]
+		
```
remove empty lines.

tchaikov commented on src/pybind/mgr/dashboard/static/AdminLTE-2.3.7/plugins/datatables/datatables_custom_darkUI.css:
```diff
@@ -1,357 +0,0 @@
-/*
```
squash this change into 65bd5188c00a8dd68823fb5db2d85c5a573e75aa. in other words, please avoid fix-the-fix changes in a single PR.

tchaikov commented on src/pybind/mgr/dashboard/module.py:
```diff
@@ -751,6 +752,133 @@ def get_perf_schema(self, **args):
+                        max_sync = sync_count
+
+                for i in range(0,len(data[ref_osd])):
+
```
remove the empty line.

tchaikov commented on src/pybind/mgr/dashboard/module.py:
```diff
@@ -751,6 +752,133 @@ def get_perf_schema(self, **args):
+                    var = sum(diff)/length
+                    stdev = math.sqrt(var)
+                    counter_stats_data["stdev"].append([tseq[0], stdev])
+                    counter_stats_data["stdevs"]["plus_onehalf"].append([tseq[0], avg+(1.5*stdev)])
```
what are these plots for: plus_onehalf, plus_half, ...?

Saumay-Agrawal replied:
These plots are to refer to the standard deviations around average, plus_onehalf = Average + (1.5 *stdev), plus_half = Average + (0.5*stdev), minus_half  = Average - (0.5*stdev) and minus_onehalf = average - (1.5*stdev). This way distribution of latency values on both sides of the average value can be seen for the performance of OSDs.

tchaikov commented on src/pybind/mgr/dashboard/module.py:
```diff
@@ -751,6 +752,133 @@ def get_perf_schema(self, **args):
+                )
+
+            @cherrypy.expose
+            @cherrypy.tools.json_out()
```
again, why shall we expose this method?

Saumay-Agrawal replied:
Again, This function is no longer there in the final commit. I would like your opinion on squashing all the changes related to graphs and TS in a single commit.

tchaikov commented on src/pybind/mgr/dashboard/module.py:
```diff
@@ -751,6 +752,133 @@ def get_perf_schema(self, **args):
+            @cherrypy.tools.json_out()
+            def get_counter_stats(self, path):
+                
+                data = self.get_counter_allosd(path)
```
could you use a better name than data? for example `counters`?

tchaikov commented on src/pybind/mgr/dashboard/module.py:
```diff
@@ -751,6 +752,133 @@ def get_perf_schema(self, **args):
+                    "osds_out_sync" : []
+                    }
+
+                result = []
```
`time_seqs`.

Saumay-Agrawal commented on src/pybind/mgr/dashboard/osd_perf.html:
```diff
@@ -121,26 +121,34 @@
+    </div>
+
+<hr>
+
```
@tchaikov  I tried to remove the empty lines, but they are not showing up here. Ironically, they are removed in my local repo and the file is committed too. Can you please help me out in this?

tchaikov commented:
@Saumay-Agrawal could you drop unrelated commits (d6d502c043d230ded31656564ee42493fa12add5 and a2a6a4d) from this PR?

tchaikov commented on src/pybind/mgr/dashboard/osd_perf_graph.html:
```diff
@@ -0,0 +1,412 @@
+</section>
+
+<section class="content">
+	
```
remove the empty lines.

tchaikov commented on src/pybind/mgr/dashboard/osd_perf_graph.html:
```diff
@@ -0,0 +1,412 @@
+{% block content %}
+
+<section class="content-header">
+    <h1>
```
nit, put this tag in a single line.

tchaikov commented on src/pybind/mgr/dashboard/osd_perf_graph.html:
```diff
@@ -0,0 +1,412 @@
+
+<section class="content">
+	
+	<h3>Read Latencies</h3>
```
please just use space for indent.

tchaikov commented on src/pybind/mgr/dashboard/osd_perf_graph.html:
```diff
@@ -0,0 +1,412 @@
+	
+	<h3>Read Latencies</h3>
+    <div class="row">
+		<div id="read1" class="col-md-6">
```
ditto.

tchaikov commented on src/pybind/mgr/dashboard/osd_perf_graph.html:
```diff
@@ -0,0 +1,412 @@
+	    });
+	}; 
+
+	var convert_timeseries = function(source_series)
```
indent.

tchaikov commented on src/pybind/mgr/dashboard/osd_perf_graph.html:
```diff
@@ -0,0 +1,412 @@
+    };
+
+    var plot_read = function() {
+
```
remove empty line.

tchaikov commented on src/pybind/mgr/dashboard/servers.html:
```diff
@@ -79,6 +82,15 @@
+        });
+      });
+    </script>
+
```
remove empty line.

nrdmn commented on src/pybind/mgr/dashboard/base.html:
```diff
@@ -31,6 +33,7 @@
     <script src="https://cdnjs.cloudflare.com/ajax/libs/moment.js/2.17.1/moment.min.js"></script>
     <script src="https://cdnjs.cloudflare.com/ajax/libs/Chart.js/2.4.0/Chart.min.js"></script>
 
+    <script src="/static/AdminLTE-2.3.7/plugins/datatables/jquery.dataTables.min.js"></script>
```
This script has already been added in de9e6a7ef0476f62ea4e1116d057e2f79089c185 but in a different line

nrdmn commented:
This PR conflicts with #17119. I've prepared a patch to resolve this.

Saumay-Agrawal commented:
@nrdmn Thanks for the patch. I've merged it and resolved the conflicts with #17119.

tchaikov commented:
@Saumay-Agrawal please don't use "merge pull request" to include @nrdmn 's pull request, you should have use "rebased and merge". normally we don't want PRs to include "merge" commit. you might want to rebase your change locally and repush.

jcsp commented on src/pybind/mgr/dashboard/base.html:
```diff
@@ -17,9 +17,9 @@
+          href="/static/AdminLTE-2.3.7/dist/css/skins/skin-blue.min.css">
     <link rel="stylesheet"
-          href="{{ url_prefix }}/static/AdminLTE-2.3.7/plugins/datatables/jquery.dataTables.css">
+          href="/static/AdminLTE-2.3.7/plugins/datatables/jquery.dataTables.css">
```
this has removed the {{url_prefix}}, it should remain (and also appear in the datatables url below)

jcsp commented:
I get this backtrace when visiting /osd_perf_graph/:
```
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/cherrypy/_cprequest.py", line 670, in respond
    response.body = self.handler()
  File "/usr/lib/python2.7/dist-packages/cherrypy/lib/encoding.py", line 217, in __call__
    self.body = self.oldhandler(*args, **kwargs)
  File "/usr/lib/python2.7/dist-packages/cherrypy/_cpdispatch.py", line 61, in __call__
    return self.callable(*self.args, **self.kwargs)
  File "/home/jspray/ceph.bravo/src/pybind/mgr/dashboard/module.py", line 953, in osd_perf_graph
    content_data = self.get_perf_graph_data()
  File "/home/jspray/ceph.bravo/src/pybind/mgr/dashboard/module.py", line 943, in get_perf_graph_data
    data = self.get_counter_stats(counter)
  File "/home/jspray/ceph.bravo/src/pybind/mgr/dashboard/module.py", line 879, in get_counter_stats
    epoch = str(counters[osd_id][0][0])
IndexError: list index out of range
```

jcsp commented:
This seems to be abandoned -- probably not sensible to merge any of it given that it presumably doesn't cope with large numbers of OSDs
